--- Makefile	2008-07-31 22:45:10.000000000 +0200
+++ Makefile	2008-07-31 22:46:57.000000000 +0200
@@ -23,12 +23,13 @@
 #
 #
 
+DESTDIR	?= /
 LIB     = -L/usr/local/lib
 INC     = -I/usr/local/include
-EXE	= /usr/local/bin
-MAN     = /usr/share/man/man1
+EXE	= $(DESTDIR)/usr/sbin
+MAN     = $(DESTDIR)/usr/share/man/man1
 CPROG	= comgt
-SCRIPTPATH = /etc/comgt/
+SCRIPTPATH = $(DESTDIR)/etc/comgt/
 SCRIPTSRC = ./scripts/
 BIN     = $(CPROG) 
 MANP	= comgt.1 sigmon.1
