--- a/util/dst-utils/dst_test.c
+++ b/util/dst-utils/dst_test.c
@@ -111,6 +111,7 @@
 	return 0;
 }

+#if defined CA_SET_PID
 static int dst_set_pid(int cafd)
 {
 	if ((ioctl(cafd, CA_SET_PID)) < 0) {
@@ -120,6 +121,7 @@

 	return 0;
 }
+#endif

 static int dst_get_descr(int cafd)
 {
@@ -230,8 +232,12 @@
 				dst_reset(cafd);
 				break;
 			case 'p':
+#if defined CA_SET_PID
 				printf("%s: PID\n", __FUNCTION__);
 				dst_set_pid(cafd);
+#else
+				printf("%s: PID not supported\n", __FUNCTION__);
+#endif
 				break;
 			case 'g':
 				printf("%s: Get Desc\n", __FUNCTION__);
--- a/util/scan/Makefile        Fri Mar 21 20:26:36 2014 +0100
+++ b/util/scan/Makefile        Sun Jan 07 22:41:44 2018 +0100
@@ -25,6 +25,6 @@
 $(objects): atsc_psip_section.c atsc_psip_section.h
 
 atsc_psip_section.c atsc_psip_section.h:
-	perl section_generate.pl atsc_psip_section.pl
+	perl section_generate.pl ./atsc_psip_section.pl
 
 include ../../Make.rules
