diff -r 8da7ade11cce drivers/mge-hid.c
--- a/drivers/mge-hid.c	Fri Jul 27 02:04:13 2018 +0300
+++ b/drivers/mge-hid.c	Fri Jul 27 02:05:44 2018 +0300
@@ -37,7 +37,7 @@
 #include "usbhid-ups.h"
 #include "mge-hid.h"
 
-#define MGE_HID_VERSION		"MGE HID 1.39"
+#define MGE_HID_VERSION		"MGE HID 1.40"
 
 /* (prev. MGE Office Protection Systems, prev. MGE UPS SYSTEMS) */
 /* Eaton */
@@ -55,6 +55,9 @@
 /* AEG */
 #define AEG_VENDORID 0x2b2d
 
+/* Phoenixtec Power Co., Ltd */
+#define PHOENIXTEC 0x06da
+
 #ifndef SHUT_MODE
 #include "usb-common.h"
 
@@ -80,6 +83,7 @@
 
 	/* PROTECT B / NAS */
 	{ USB_DEVICE(AEG_VENDORID, 0xffff), NULL },
+	{ USB_DEVICE(PHOENIXTEC, 0xffff), NULL },
 
 	/* Terminating entry */
 	{ -1, -1, NULL }
