diff --git a/src/third_party/s2/hash.h b/src/third_party/s2/hash.h
index 48c6cb42..bdc92b99 100644
--- a/src/third_party/s2/hash.h
+++ b/src/third_party/s2/hash.h
@@ -2,10 +2,12 @@
 #define THIRD_PARTY_S2_HASH_H_
 
 #include <unordered_map>
-#define hash_map std::unordered_map
+//#define hash_map std::unordered_map
+template<class _Key, class _Tp, class _Hash = std::hash<_Key>, class _Pred = std::equal_to<_Key>, class _Alloc = std::allocator< std::pair<const _Key, _Tp> >> using hash_map = std::unordered_map<_Key, _Tp, _Hash, _Pred, _Alloc>;
 
 #include <unordered_set>
-#define hash_set std::unordered_set
+//#define hash_set std::unordered_set
+template<class _Value, class _Hash = std::hash<_Value>, class _Pred = std::equal_to<_Value>, class _Alloc = std::allocator<_Value>> using hash_set = std::unordered_set<_Value, _Hash, _Pred, _Alloc>;
 
 #define HASH_NAMESPACE_START namespace std {
 #define HASH_NAMESPACE_END }
