diff --git a/SConstruct b/SConstruct
index 07579349..7236c6c2 100644
--- a/SConstruct
+++ b/SConstruct
@@ -1477,6 +1477,12 @@ env_vars.Add(
     default=mongo_platform.get_running_os_name(),
 )
 
+env_vars.Add(
+    'DISABLE_AVX',
+    help='Disable avx',
+    default='no'
+)
+
 env_vars.Add(
     'TOOLS',
     help='Sets the list of SCons tools to add to the environment',
diff --git a/src/third_party/mozjs/SConscript b/src/third_party/mozjs/SConscript
index 7d7e132d..35ec5aae 100644
--- a/src/third_party/mozjs/SConscript
+++ b/src/third_party/mozjs/SConscript
@@ -145,7 +145,8 @@ sources = [
 ]
 
 if env['TARGET_ARCH'] == 'x86_64' and not env.TargetOSIs('windows'):
-    env.Append(CCFLAGS=['-mavx2'])
+    if env['DISABLE_AVX'] != 'yes':
+        env.Append(CCFLAGS=['-mavx2'])
     sources.extend(["extract/mozglue/misc/SIMD_avx2.cpp", "extract/mozglue/misc/SSE.cpp"])
 
 if env.TargetOSIs('windows'):
diff --git a/src/third_party/mozjs/extract/mozglue/misc/SIMD_avx2.cpp b/src/third_party/mozjs/extract/mozglue/misc/SIMD_avx2.cpp
index a1467c7a..9a714419 100644
--- a/src/third_party/mozjs/extract/mozglue/misc/SIMD_avx2.cpp
+++ b/src/third_party/mozjs/extract/mozglue/misc/SIMD_avx2.cpp
@@ -11,7 +11,7 @@
 // Restricting to x86_64 simplifies things, and we're not particularly
 // worried about slightly degraded performance on 32 bit processors which
 // support AVX2, as this should be quite a minority.
-#if defined(MOZILLA_MAY_SUPPORT_AVX2) && defined(__x86_64__)
+#if defined(MOZILLA_MAY_SUPPORT_AVX2) && defined(__x86_64__) && !defined(DISABLE_AVX)
 
 #  include <cstring>
 #  include <immintrin.h>
