diff --git a/src/third_party/mozjs/SConscript b/src/third_party/mozjs/SConscript
index 87b09502..9354dc16 100644
--- a/src/third_party/mozjs/SConscript
+++ b/src/third_party/mozjs/SConscript
@@ -148,8 +148,10 @@ sources = [
 ]
 
 if env['TARGET_ARCH'] == 'x86_64' and not env.TargetOSIs('windows'):
-    env.Append(CCFLAGS=['-mavx2'])
-    sources.extend(["extract/mozglue/misc/SIMD_avx2.cpp", "extract/mozglue/misc/SSE.cpp"])
+    ccflags_str = str(env.get('CCFLAGS', ''))
+    if '-mno-avx' not in ccflags_str:
+        env.Append(CCFLAGS=['-mavx2'])
+        sources.extend(["extract/mozglue/misc/SIMD_avx2.cpp", "extract/mozglue/misc/SSE.cpp"])
 
 if env.TargetOSIs('windows'):
     sources.extend([
